File: Algorithm/BinarySearch/leetcode/35. Search Insert Position/Search-Insert-Position-GPT.py
# ```python
# Runtime
# 0
# ms
# Beats
# 100.00%
# Memory
# 18.33
# MB
# Beats
# 72.33%

from typing import List

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        """
        パラメーター:
            nums: List[int] - 昇順にソートされた整数配列 (長さ1〜10^4, 要素は-10^4〜10^4)
            target: int - 探索する整数 (-10^4〜10^4)
        返却値:
            int - targetが存在する場合はそのインデックス、存在しない場合は挿入すべきインデックス

        計算量:
            時間計算量 O(log n)  (二分探索)
            空間計算量 O(1)      (追加メモリはほぼ定数)

        実行環境:
            Python (CPython 3.11.4), LeetCodeで動作確認済み
            Pylanceで型エラーなし
        """
        left: int = 0
        right: int = len(nums) - 1

        while left <= right:
            # オーバーフロー防止のため (left + right)//2 ではなく以下でも可
            mid: int = left + (right - left) // 2

            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1  # 右側探索
            else:
                right = mid - 1 # 左側探索

        # 見つからない場合は挿入位置がleft
        return left
# ```

# ---

# ### 実装上のポイント

# 1. **二分探索**を使用し、最大でも `log2(n)` 回のループ

#    * `n=10^4` の場合でも最大約 14 回で高速
# 2. **追加メモリ不要**で O(1)
# 3. **Pylance 型推論対応**

#    * `List[int]` や `int` を明示して型エラーを防止
# 4. **LeetCode にそのまま提出可能**

# ---

# もし希望なら、この Python 実装の **二分探索の動作過程を図解** したバージョンも作成できますが、作成しますか？

File: Algorithm/BinarySearch/leetcode/35. Search Insert Position/Search-Insert-Position-GPT.ts
// ```typescript
// Runtime
// 0
// ms
// Beats
// 100.00%
// Memory
// 55.79
// MB
// Beats
// 37.52%
/**
 * 関数名: searchInsert
 * パラメーター:
 *   nums: number[] - 昇順にソートされた整数配列 (長さ1〜10^4, 要素は-10^4〜10^4)
 *   target: number - 探索する整数 (-10^4〜10^4)
 * 返却値:
 *   number - targetが存在する場合はそのインデックス、存在しない場合は挿入すべきインデックス
 *
 * 計算量: O(log n)  (二分探索)
 * メモリ消費量: O(1)  (追加メモリはほぼ定数)
 */
function searchInsertGPT(nums: number[], target: number): number {
    let left = 0;
    let right = nums.length - 1;

    while (left <= right) {
        // 安全に中央インデックスを求める
        const mid = left + ((right - left) >> 1);

        if (nums[mid] === target) {
            return mid; // targetが見つかった場合
        } else if (nums[mid] < target) {
            left = mid + 1; // 右半分を探索
        } else {
            right = mid - 1; // 左半分を探索
        }
    }

    // targetが存在しない場合、挿入位置はleft
    return left;
}
// ```

// ### 実装上のポイント

// * **二分探索**により、配列長 `n` に対して計算量は **O(log n)**
// * **追加メモリ不要**で、メモリ消費は **O(1)**
// * `mid` 計算は `left + ((right - left) >> 1)` にすることで高速かつオーバーフロー防止

// このまま TypeScript (Node.js 18.16.1, LeetCode) に提出可能です。
